packages/eslint-plugin/src/rules/ban-imports.js

```diff
@@ -69,49 +90,71 @@ module.exports = createRule({
       return {
         path: entry.path,
         pathRegex: entry.pathRegex ? new RegExp(entry.pathRegex) : undefined,
-        names: entry.names,
+        names: entry.names
+          ? entry.names.map(name => (typeof name === 'string' ? name : new RegExp(name.regex)))
+          : undefined,
       };
     });
 
-    return {
-      ImportDeclaration: imprt => {
-        if (imprt.source.type !== AST_NODE_TYPES.Literal || typeof imprt.source.value !== 'string') {
-          return;
-        }
+    /**
+     * @param {ImportDeclaration | ExportNamedDeclaration} importOrExport
+     */
+    function checkImportOrExport(importOrExport) {
+      if (
+        !importOrExport.source ||
+        importOrExport.source.type !== AST_NODE_TYPES.Literal ||
+        typeof importOrExport.source.value !== 'string'
+      ) {
+        return;
+      }
 
-        const importPath = imprt.source.value;
-        // Filter out default imports and namespace (star) imports
-        const specifiers = /** @type {ImportSpecifier[]} */ (imprt.specifiers.filter(
-          spec => spec.type === AST_NODE_TYPES.ImportSpecifier,
-        ));
+      const importPath = importOrExport.source.value;
+      const specifiers =
+        importOrExport.type === AST_NODE_TYPES.ExportNamedDeclaration
+          ? importOrExport.specifiers
+          : // Filter out default imports and namespace (star) imports
+            /** @type {ImportSpecifier[]} */ (importOrExport.specifiers.filter(
+              spec => spec.type === AST_NODE_TYPES.ImportSpecifier,
+            ));
 
-        for (const rule of options) {
-          const { path, pathRegex, names, message = '' } = rule;
-          const pathForErrors = path || /** @type {RegExp} */ (pathRegex).source;
-          const messageForErrors = message && `: ${message}`;
+      for (const rule of options) {
+        const { path, pathRegex, names, message = '' } = rule;
+        const pathForErrors = path || /** @type {RegExp} */ (pathRegex).source;
```

Comment:
Based on earlier validation, we know that exactly one of `path` or `pathRegex` is defined (so in TS the cast would have been a non-null assertion). But I realized it would make more sense to modify the options validation/mapping.